okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/UrlSelectorImpl.java

```diff
@@ -113,84 +115,102 @@ private static String switchWsToHttp(String url) {
     }
 
+    /**
+     * Rewrites the current URL to use the new {@code redirectBaseUrl}, if the path prefix is compatible, otherwise
+     * it returns {@link Optional#empty()}.
+     *
+     * Also updates the {@link #lastBaseUrl} with the given {@code redirectBaseUrl}.
+     *
+     * @param redirectBaseUrl  expected to be an actual base url that exists in {@link #baseUrls}.
+     */
     private Optional<HttpUrl> redirectTo(HttpUrl current, HttpUrl redirectBaseUrl) {
-        Optional<Integer> baseUrlIndex = indexFor(redirectBaseUrl);
-        baseUrlIndex.ifPresent(currentUrl::set);
-
-        return baseUrlIndex
-                .map(baseUrls.get()::get)
-                .flatMap(baseUrl -> {
-                    if (!isPathPrefixFor(baseUrl, current)) {
-                        // The requested redirectBaseUrl has a path that is not compatible with
-                        // the path of the current URL
-                        return Optional.empty();
-                    } else {
-                        return Optional.of(current.newBuilder()
-                                .scheme(baseUrl.scheme())
-                                .host(baseUrl.host())
-                                .port(baseUrl.port())
-                                .encodedPath(
-                                        baseUrl.encodedPath()  // matching prefix
-                                                + current.encodedPath().substring(baseUrl.encodedPath().length()))
-                                .build());
-                    }
-                });
+        lastBaseUrl.set(redirectBaseUrl);
+
+        if (!isPathPrefixFor(redirectBaseUrl, current)) {
+            // The requested redirectBaseUrl has a path that is not compatible with
+            // the path of the current URL
+            return Optional.empty();
+        } else {
+            return Optional.of(current.newBuilder()
+                    .scheme(redirectBaseUrl.scheme())
+                    .host(redirectBaseUrl.host())
+                    .port(redirectBaseUrl.port())
+                    .encodedPath(
+                            redirectBaseUrl.encodedPath()  // matching prefix
+                                    + current.encodedPath().substring(redirectBaseUrl.encodedPath().length()))
+                    .build());
+        }
     }
 
     @Override
     public Optional<HttpUrl> redirectToNext(HttpUrl existingUrl) {
+        List<HttpUrl> httpUrls = baseUrls.get();
         // if possible, determine the index of the passed in url (so we can be sure to return a url which is different)
-        Optional<Integer> existingUrlIndex = indexFor(existingUrl);
-
-        int potentialNextIndex = existingUrlIndex.orElse(currentUrl.get());
+        Optional<Integer> existingUrlIndex = indexFor(existingUrl, httpUrls);
+        int thisIndex = existingUrlIndex.orElseGet(() -> getLastIndex(httpUrls));
```

Comment:
there has to be a better name than `thisIndex`. previouslyUsedIndex? startIndex?

Reply:
I thought of startIndex too, we can use that I guess! it's not necessarily previouslyUsed cos it preferrentially comes from the existingUrl rather than the lastUsedUrl